### PR TITLE
Use selection buffer in ray question (ogre2) - part 2

### PR DESCRIPTION
**Pull request by iche033**
# 🎉 New feature

* follow-up to pull request #378 
* requires https://github.com/ignitionrobotics/ign-gazebo/pull/969

## Summary

This further improves performance of ray queries. As described in #378, there are 2 stages in our ray query implementation. This PR focuses on stage 2. The selection buffer implementation in ogre2 is extended to return depth data which is then used compute the intersection point of user mouse clicks. This replaces the slow CPU based ray-triangle intersection tests in stage 2 of the current ray query implementation.

Note the ray-triangle intersection code is not removed, and can still be used if a selection buffer is not available.

## Test it

see  https://github.com/ignitionrobotics/ign-gazebo/pull/969

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: ogre2/src/Ogre2RayQuery.cc [hunk 1]
```diff
@@ -33,11 +33,11 @@ class ignition::rendering::Ogre2RayQueryPrivate
...
 
-  //// \brief Pointer to camera
-  public: Ogre2CameraPtr camera;
-
-  /// Image pos to cast the ray from
+  /// \brief Image pos to cast the ray from
   public: math::Vector2i imgPos;
+
+  /// \brief Pointer to camera used for ray query
+  public: Ogre2CameraPtr camera;
```
Comment (ahcorde) [inline ogre2/src/Ogre2RayQuery.cc:40]:
```suggestion
  public: Ogre2CameraPtr camera{nullptr};
```

File: ogre2/src/Ogre2SelectionBuffer.cc [hunk 2]
```diff
@@ -63,10 +64,13 @@ class ignition::rendering::Ogre2SelectionBufferPrivate
...
 
   /// \brief Render texture data buffer
-  public: uint8_t *buffer = nullptr;
+  public: float *buffer = nullptr;
 
   /// \brief Ogre pixel box that contains description of the data buffer
   public: Ogre::PixelBox *pixelBox = nullptr;
+
+  /// \brief The selection buffer material
+  public: Ogre::MaterialPtr selectionMaterial;
```
Comment (ahcorde) [inline ogre2/src/Ogre2SelectionBuffer.cc:73]:
```suggestion
  public: Ogre::MaterialPtr selectionMaterial{nullptr};
```

File: ogre2/src/Ogre2SelectionBuffer.cc [hunk 3]
```diff
@@ -242,31 +413,51 @@ Ogre::Item *Ogre2SelectionBuffer::OnSelectionClick(const int _x, const int _y)
...
+      this->dataPtr->camera->getParentSceneNode()->_getDerivedPosition());
+  math::Pose3d p(pos, rot);
+  point = rot * point + pos;
+
   ignition::math::Color cv;
-  cv.SetFromARGB(color);
   cv.A(1.0);
+  cv.R(r/255.0);
+  cv.G(g/255.0);
+  cv.B(b/255.0);
```
Comment (ahcorde) [inline ogre2/src/Ogre2SelectionBuffer.cc:441]:
```suggestion
  cv.R(r / 255.0);
  cv.G(g / 255.0);
  cv.B(b / 255.0);
```

File: ogre2/src/Ogre2SelectionBuffer.cc [hunk 4]
```diff
@@ -161,55 +210,157 @@ void Ogre2SelectionBuffer::CreateRTTBuffer()
...
-  auto pass = nodeSeq[0]->_getPasses()[1]->getDefinition();
-  auto scenePass = dynamic_cast<const Ogre::CompositorPassSceneDef *>(pass);
-  const_cast<Ogre::CompositorPassSceneDef *>(scenePass)->mVisibilityMask =
-      IGN_VISIBILITY_SELECTABLE;
+  // add the listener
+  Ogre::CompositorNode *node =
+      this->dataPtr->ogreCompositorWorkspace->getNodeSequence()[0];
+  auto channelsTex = node->getLocalTextures();
+
+  for (auto c : channelsTex)
```
Comment (ahcorde) [inline ogre2/src/Ogre2SelectionBuffer.cc:309]:
```suggestion
  for (auto & c : channelsTex)
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: ogre2/src/Ogre2RayQuery.cc (hunk 1 above)
Comment (iche033) [reply ogre2/src/Ogre2RayQuery.cc]: 58054d144ea7100363cfb5dcb036d23625c4f7f3

File: ogre2/src/Ogre2SelectionBuffer.cc (hunk 3 above)
Comment (iche033) [reply ogre2/src/Ogre2SelectionBuffer.cc]: 58054d144ea7100363cfb5dcb036d23625c4f7f3

File: ogre2/src/Ogre2SelectionBuffer.cc (hunk 4 above)
Comment (iche033) [reply ogre2/src/Ogre2SelectionBuffer.cc]: 58054d144ea7100363cfb5dcb036d23625c4f7f3

File: ogre2/src/Ogre2SelectionBuffer.cc (hunk 2 above)
Comment (iche033) [reply ogre2/src/Ogre2SelectionBuffer.cc:73]: looks like `Ogre::SharedPtr` can't be set or init to `nullptr`.

**Review by nkoenig — APPROVED** [review]

**Pull request merged**